cpp/ql/lib/semmle/code/cpp/models/implementations/Gets.qll

```diff
@@ -54,17 +53,60 @@ private class GetsFunction extends DataFlowFunction, TaintFunction, ArrayFunctio
   }
 
   override predicate hasArrayWithVariableSize(int bufParam, int countParam) {
-    not this.hasName("gets") and
     bufParam = 0 and
     countParam = 1
   }
 
+  override predicate hasArrayOutput(int bufParam) { bufParam = 0 }
+
+  override predicate hasSocketInput(FunctionInput input) { input.isParameter(2) }
```

Comment:
This isn't really something we need to fix in this PR, but should this really be `isParameter(2)` and not `isParameterDeref(2)`? Presumably, `isParameter(2)` should only hold if `fgets` had a prototype like
```cpp
char *fgets(char *str, int n, FILE stream)
```
(i.e., taking the `FILE` argument by value).

I see that all of our implementations of `hasSocketInput` uses `isParameter`, so it's probably not something we should change in this PR.

Reply:
I don't think it matters as long as we're consistent.  Note that `hasSocketInput` is not describing the data coming from the socket, just the parameter the socket is specified by (though perhaps the QLDoc could be clearer on this).

Reply:
True. We lose the option to distinguish between functions that take a socket inputs by value and by functions that take socket inputs by pointer or reference, though. So if we need that distinction at some point we need to go back and update the implementations to correctly reflect this.

Reply:
Good point.

Reply:
Fixed.  In fact most cases are correctly `isParameter`, because sockets (as opposed to streams) are expressed as a simple `int`, e.g.
```
ssize_t recv(int sockfd, void *buf, size_t len, int flags);
```

Reply:
Good point. Thanks for looking into it!